File: 250514/최고의 13위치 2/best-place-of-13-2.java
import java.util.Scanner;
public class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        int[][] arr = new int[n][n];
        for(int i = 0; i < n; i++)
            for(int j = 0; j < n; j++)
                arr[i][j] = sc.nextInt();
        // Please write your code here.

        int maxCoin = -1;
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n-2; j++) {
                for (int k = i; k < n; k++) {
                    boolean jAfterButton = false;
                    for (int l = 0; l < n-2; l++) {
                        if (k==i && !jAfterButton) {
                            jAfterButton = true;
                            l = j+3;
                            if (l >= n-2) {
                                break;
                            }
                        }
                        maxCoin = Math.max(maxCoin, arr[i][j] + arr[i][j+1] + arr[i][j+2] + arr[k][l] + arr[k][l+1] + arr[k][l+2]);
                    }
                }
            }
        }

        System.out.print(maxCoin);
    }
}

// 

File: 251022/최고의 13위치 2/best-place-of-13-2.py
n = int(input())
arr = [list(map(int, input().split())) for _ in range(n)]

# Please write your code here.

max_coins = 0

for i in range(n) :
    for j in range(n-2) :


        for a in range(i,n) :
            if a == i :
                for b in range(j+3, n-2) :
                    max_coins = max(max_coins, arr[i][j] + arr[i][j+1] + arr[i][j+2] + arr[a][b] + arr[a][b+1] + arr[a][b+2])

            else :
                for b in range(n-2) :
                    max_coins = max(max_coins, arr[i][j] + arr[i][j+1] + arr[i][j+2] + arr[a][b] + arr[a][b+1] + arr[a][b+2])

print(max_coins)
